keps/sig-api-machinery/xxxx-progress-notify/README.md

```diff
@@ -0,0 +1,456 @@
+and kube-apiserver replicas, kube-apiserver contains an in-memory cache
+called `watch cache` from which all watch requests as well as non-quorum
+reads (in opt-in fashion) are served. Watch cache is being propagated
+directly from etcd using typical "list + watch" pattern and stores both
+the current state as well as som recent 'transaction log'. There is a
+separate watch cache for each resource type version.
+
+Digging into more details, the way how watch cache is propagate is:
+- on start, quorum list is send to etcd
+- the watch cache is set to be synced to resource version of that list
+- from that point, watch to etcd is established and watch cache is
+  updated by incoming watch event stream; note that only objects of
+  a given resource type are watched
+- if watch cannot be reestablished, all watchers connected to it are
+  disconnected, a new quorum list is send and watchcache is reset to
+  that point in time
+
+It's important to note that while only objects of a given resource type
+are watched (thus resource version to which watch cache is synced can
+only be updated to a value that some object of that type had), the quorum
+list, even though it is only requesting objects of that resource type,
+returns "globally current version for the whole etcd".
+
+This is resulting in certain problems described above that this KEP is
+trying to address.
+
+## Motivation
+
+While the inability to reestablish watch doesn't happen often in practice,
+the kube-apiserver restarts obviously does (e.g. on upgrades). And this
+is resulting in two main problems:
+
+1. In case of rolling upgrade of kube-apiserver, when no object of a given
+   resource type is changing (but objects of other types do), all watchers
+   will eventually be forced to relist, causing significant performance
+   and scalability issues for larger clusters. See the example below for
+   clarification.
+
+1. If no object of a given resource type is changing after kube-apiserver
+  startup, watch cache may stuck being synced to resource version not being
+	version of any object of a given type for extended period of time. See
+  https://github.com/kubernetes/kubernetes/issues/91073 for more details.
+
+To illustrate the problems better, imagine the following example:
+- there are two kube-apiserver, we are performing their rolling upgrade
+- there are only two resource types: X and Y (for simplicity)
+- there is one object of type X: x with rv=100
+- there is one object of type Y: y with rv=101
+- there is a watch W for objects of type X already synced to rv=100 and
+  attched to kube-apiserver-1
+- both kube-apiserver are up-to-date with watchcache for X synced to rv=100
+
+- T1: kube-apiserver-1 is being restarted
+- T2: watch W is being reestablished from rv=100 from kube-apiserver-2
+- T3: kube-apiserver-1 is initialized, with watcache X synced to rv=101
+- T4: y is being updated to rv=102
+- T5: kube-apiserver-2 is being restarted
```

Comment:
I think you don't need multiple apiservers to illustrate the problem. Actually the problem is most acute when there's just a single apiserver.

If there were 100 apiservers, this wouldn't be nearly as bad of a problem, because by the time you reboot the final one, there's lots of others with sufficient history in their windows. (Of course, the change rate matters too, but in low change rate systems (which cause more out of window errors) it's also less problematic to do full re lists, as they're less busy.)

So, adding apiservers actually makes it a little better, but just going to 3 or 5 is not sufficient to fix the problem.

It might be clearer to demonstrate the problem with a single apiserver and then explain why adding a few apiservers doesn't fix it.